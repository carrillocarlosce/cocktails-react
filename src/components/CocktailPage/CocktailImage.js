import React from "react";
import { withStyles } from "@material-ui/core/styles";

const styles = {
  image: {
    boxShadow: "0 19px 38px rgba(0,0,0,0.30), 0 15px 12px rgba(0,0,0,0.22)",
    padding: ".3em",
    marginRight: "2em"
  }
};

const CocktailImage = ({ name, image, classes }) => (
  <img width="100%" className={classes.image} src={image} alt={{ name }} />
);

export default withStyles(styles)(CocktailImage);
