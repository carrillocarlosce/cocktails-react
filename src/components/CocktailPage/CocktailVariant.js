import React from "react";
import Card from "@material-ui/core/Card";
import CardHeader from "@material-ui/core/CardHeader";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import LocalBar from "@material-ui/icons/LocalBar";
import Typography from "@material-ui/core/Typography";
import { withStyles } from "@material-ui/core/styles";

const styles = {
  circle: {
    width: ".8em",
    height: ".8em",
    display: "inline-block",
    float: "right"
  },
  card: {
    width: "30%",
    margin: ".5em"
  },
  cardContent: {
    padding: "1em"
  },
  button: {
    padding: "0.5em 1em"
  },
  title: {
    fontSize: 20,
    marginTop: 0,
    marginBottom: 0
  },
  subHeader: {
    fontSize: 14,
    fontStyle: "italic"
  },
  category: {
    fontSize: 12
  },
  media: {
    height: 0,
    paddingTop: "56.25%" // 16:9
  },
  prep: {
    fontStyle: "italic"
  },
  glass: {}
};

const CocktailVariant = ({
  cocktail: { name, image, category, glass, preparation, ingredients } = {},
  classes
}) => {
  return (
    <Card className={classes.card}>
      <CardHeader
        title={<h1 className={classes.title}>{name}</h1>}
        subheader={<span className={classes.subHeader}>{category}</span>}
      />
      <CardMedia className={classes.media} image={image} title={name} />

      <CardContent className={classes.cardContent}>
        <ul>
          {ingredients &&
            ingredients.map((item, idx) => (
              <li key={idx}>
                <Typography className={classes.ingredients}>{item}</Typography>
              </li>
            ))}
        </ul>
        <br />
        <Typography component="p" className={classes.prep}>
          {preparation}
        </Typography>
        <br />
        {glass && (
          <Typography component="p" color="textSecondary">
            <LocalBar fontSize="inherit" />
            &nbsp;
            {glass}
          </Typography>
        )}
      </CardContent>
    </Card>
  );
};

export default withStyles(styles)(CocktailVariant);
