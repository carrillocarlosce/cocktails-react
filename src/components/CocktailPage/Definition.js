import React from "react";
import { withStyles } from "@material-ui/core";

const styles = theme => ({
  dt: { fontWeight: "bold" },
  dd: { marginLeft: 0, marginBottom: ".5em" }
});

const Definition = ({ title, description, classes }) => {
  if (!description) return null;
  return (
    <>
      <dt className={classes.dt}>{title}</dt>
      <dd className={classes.dd}>{description}</dd>
    </>
  );
};

export default withStyles(styles)(Definition);
