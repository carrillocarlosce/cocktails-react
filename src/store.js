import appReducer from "./reducers";
import { createStore, applyMiddleware, compose } from "redux";
import {
  fetchCocktails,
  fetchIngredients,
  fetchGlasses
} from "./services/cocktail.service";
import { loadCocktails, loadIngredients, loadGlasses } from "./actions";
import {
  persistCurrentState,
  supportsPersistence
} from "./utilities/persistence";
import throttle from "lodash/throttle";

import thunk from "redux-thunk";

const devtools =
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__();

const middlewares = compose(
  applyMiddleware(thunk),
  devtools || (a => a)
);

const store = createStore(appReducer, middlewares);

// persist parts of the store whenever it changes.
if (supportsPersistence()) {
  store.subscribe(
    throttle(() => {
      persistCurrentState(store.getState(), ["bar", "settings", "favourites"]);
    })
  );
}

// Load glasses, ingredients and cocktails (order is important - cocktails
// needs to be loaded last as the views all assume ingredietns and glasses
// will be available to query)

// TODO: This could be a lot nicer - there is another ticket to
// look into using await/async syntax.
fetchGlasses()
  .then(function(glasses) {
    store.dispatch(loadGlasses(glasses));
  })
  .then(fetchIngredients)
  .then(function(ingredients) {
    store.dispatch(loadIngredients(ingredients));
  })
  .then(fetchCocktails)
  .then(function(cocktails) {
    store.dispatch(loadCocktails(cocktails));
  });

export default store;
