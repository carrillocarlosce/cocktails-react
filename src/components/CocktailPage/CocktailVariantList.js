import React from "react";
import { Typography, GridList } from "@material-ui/core";
import { withStyles } from "@material-ui/core/styles";
import CocktailVariant from "./CocktailVariant";

const styles = theme => ({
  gridList: {
    justifyContent: "center"
  }
});

const CocktailPage = ({ cocktail, classes }) => {
  const { enrichment, enriched } = cocktail;

  if (!enriched || !enrichment.variants || !enrichment.variants.length)
    return null;

  return (
    <>
      <Typography variant="h4" gutterBottom>
        Variants
      </Typography>

      <GridList className={classes.gridList}>
        {enrichment.variants.map(variant => {
          return <CocktailVariant key={variant.name} cocktail={variant} />;
        })}
      </GridList>
    </>
  );
};

export default withStyles(styles)(CocktailPage);
