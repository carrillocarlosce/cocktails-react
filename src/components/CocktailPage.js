import React from "react";
import { connect } from "react-redux";
import {
  Typography,
  Paper,
  CircularProgress,
  GridList,
  Grid
} from "@material-ui/core";
import IngredientDetail from "./IngredientDetail";
import Definition from "./CocktailPage/Definition";
import { currentCocktailSelector } from "../selectors";
import { withStyles } from "@material-ui/core/styles";
import { bindActionCreators } from "redux";
import { enrichCocktail } from "../actions";
import CocktailVariant from "./CocktailPage/CocktailVariant";
import CocktailImage from "./CocktailPage/CocktailImage";
const styles = theme => ({
  paper: {
    padding: "1em"
  },
  root: {
    ...theme.mixins.gutters,
    justifyContent: "center"
  },
  gridList: {
    justifyContent: "center"
  },
  definitions: {
    marginTop: "1em"
  },
  progress: {
    width: "100%"
  }
});

const CocktailPage = ({ cocktail, enrichCocktail, classes, match }) => {
  if (!cocktail) return null;

  const { enriched, enriching, enrichmentFailed } = cocktail;

  if (!enriching && !enriched && !enrichmentFailed) enrichCocktail(cocktail);

  const {
    name,
    ingredients,
    preparation,
    category,
    glass,
    garnish,
    enrichment
  } = cocktail;

  return (
    <div className={classes.root}>
      <Paper className={classes.paper} square>
        <Grid container className={classes.innerContainer} spacing={10}>
          <Grid item md={8} xs={12}>
            <Typography variant="h3" gutterBottom>
              {name}
            </Typography>
            <Typography component="ul" gutterBottom>
              <>
                {ingredients.map((ingredient, idx) => {
                  return (
                    <li key={`ingredient-${idx}`}>
                      <IngredientDetail item={ingredient} />
                    </li>
                  );
                })}
              </>
            </Typography>
            <Typography
              className={classes.definitions}
              component="dl"
              gutterBottom
            >
              <>
                <Definition title="Category" description={category} />
                <Definition title="Glass" description={glass} />
                <Definition title="Preparation" description={preparation} />
                <Definition title="Garnish" description={garnish} />
                {enriched && enrichment.ibaCategory && (
                  <Definition
                    title="IBA Category"
                    description={enrichment.ibaCategory}
                  />
                )}
              </>
            </Typography>
          </Grid>
          <Grid item md={4} xs={12}>
            <div style={{ textAlign: "center" }}>
              {enriching && (
                <CircularProgress size="50%" className={classes.progress} />
              )}
              {enriched && enrichment.image && (
                <div style={{ marginRight: "1em" }}>
                  <CocktailImage image={enrichment.image} name={name} />
                </div>
              )}
            </div>
          </Grid>
        </Grid>
        <br />

        {enriched && enrichment.variants && enrichment.variants.length > 0 && (
          <>
            <Typography variant="h4" gutterBottom>
              Variants
            </Typography>

            <GridList className={classes.gridList}>
              {enrichment.variants.map(variant => {
                return (
                  <CocktailVariant key={variant.name} cocktail={variant} />
                );
              })}
            </GridList>
          </>
        )}
      </Paper>
    </div>
  );
};

const mapStateToProps = (state, ownProps) => ({
  cocktail: currentCocktailSelector(state, ownProps)
});

const mapDispatchToProps = dispatch => ({
  enrichCocktail: bindActionCreators(enrichCocktail, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(CocktailPage));
