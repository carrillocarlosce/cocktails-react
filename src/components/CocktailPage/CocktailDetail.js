import React from "react";
import { Typography } from "@material-ui/core";
import IngredientDetail from "../IngredientDetail";
import Definition from "./Definition";
import { withStyles } from "@material-ui/core/styles";

const styles = theme => ({
  definitions: {
    marginTop: "1em"
  }
});

const CocktailPage = ({ cocktail, classes }) => {
  const {
    name,
    ingredients,
    preparation,
    category,
    glass,
    garnish,
    enrichment,
    enriched
  } = cocktail;

  return (
    <>
      <Typography variant="h3" gutterBottom>
        {name}
      </Typography>
      <Typography component="ul" gutterBottom>
        <>
          {ingredients.map((ingredient, idx) => {
            return (
              <li key={`ingredient-${idx}`}>
                <IngredientDetail item={ingredient} />
              </li>
            );
          })}
        </>
      </Typography>
      <Typography className={classes.definitions} component="dl" gutterBottom>
        <>
          <Definition title="Category" description={category} />
          <Definition title="Glass" description={glass} />
          <Definition title="Preparation" description={preparation} />
          <Definition title="Garnish" description={garnish} />
          {enriched && enrichment.ibaCategory && (
            <Definition
              title="IBA Category"
              description={enrichment.ibaCategory}
            />
          )}
        </>
      </Typography>
    </>
  );
};

export default withStyles(styles)(CocktailPage);
