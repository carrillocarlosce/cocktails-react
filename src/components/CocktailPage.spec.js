import React from "react";
import renderer from "react-test-renderer";
import CocktailPage from "./CocktailPage";
import store from "../store";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { noop } from "lodash";
import cocktails from "../data/cocktails.json";

it("does not explode when rendered", () => {
  // jsdom does not implement scrollTo so we need to mock it.
  const jsdomAlert = window.scrollTo;
  window.scrollTo = noop;

  const tree = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <CocktailPage cocktail={cocktails[0]} enrichCocktail={noop} />
      </MemoryRouter>
    </Provider>
  );

  window.scrollTo = jsdomAlert;
  expect(tree).toMatchSnapshot();
});
