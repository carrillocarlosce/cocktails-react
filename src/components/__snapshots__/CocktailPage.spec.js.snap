// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not explode when rendered 1`] = `
Array [
  <div
    className="MuiBox-root MuiBox-root-6 CocktailPage-mobileImage-4"
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  />,
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root CocktailPage-cocktailDetail-1 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
    >
      <div
        className="CocktailPage-cocktailDetailContent-2"
      >
        <h1
          className="MuiTypography-root CocktailPage-cocktailTitle-112 MuiTypography-h1"
        >
          Vesper
        </h1>
        <div
          className="CocktailPage-definitions-110"
        >
          <div
            className="MuiChip-root Definition-chip-143 MuiChip-colorSecondary"
            onKeyDown={[Function]}
            onKeyUp={[Function]}
          >
            <span
              className="MuiChip-label"
            >
              Before Dinner Cocktail
            </span>
          </div>
          <div
            className="MuiChip-root Definition-chip-143 MuiChip-colorSecondary"
            onKeyDown={[Function]}
            onKeyUp={[Function]}
          >
            <span
              className="MuiChip-label"
            >
              Martini
            </span>
          </div>
          <div
            className="MuiChip-root Definition-chip-143 MuiChip-colorSecondary"
            onKeyDown={[Function]}
            onKeyUp={[Function]}
          >
            <span
              className="MuiChip-label"
            >
              Lemon twist
            </span>
          </div>
        </div>
        <div
          className="MuiPaper-root MuiPaper-elevation1 CocktailPage-cocktailDetailPaper-111 MuiPaper-rounded"
        >
          <ul
            className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
          >
            <li>
              <span>
                6 cl
                 
                <strong
                  aria-describedby={null}
                  className="Connect(IngredientDetail)-tooltip-201"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                  title="40% abv"
                >
                  Gin
                </strong>
              </span>
            </li>
            <li>
              <span>
                1.5 cl
                 
                <strong
                  aria-describedby={null}
                  className="Connect(IngredientDetail)-tooltip-201"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                  title="40% abv"
                >
                  Vodka
                </strong>
              </span>
            </li>
            <li>
              <span>
                0.75 cl
                 
                <strong
                  aria-describedby={null}
                  className="Connect(IngredientDetail)-tooltip-201"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                  title="15% abv"
                >
                  Lillet Blonde
                </strong>
              </span>
            </li>
          </ul>
          <br />
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            Shake and strain into a chilled cocktail glass.
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6"
    >
      <div
        className="CocktailPage-cocktailImage-3"
        style={
          Object {
            "backgroundImage": "url(undefined)",
            "opacity": 0,
            "visibility": "hidden",
          }
        }
      />
    </div>
  </div>,
]
`;
